### PR TITLE
Atom renderer automation strategy document

### PR DESCRIPTION
**Pull request by smurly**
initial push without image files
Signed-off-by: Scott Murray <scottmur@amazon.com>

### CODE REVIEW

File: testing-guidance/Atom-Automation-Strategy.md [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+
+This section provides an overview of our prioritization logic based on the [Test Automation Pyramid](#AtomAutomationStrategy-TestAutomationPyramid) methodology.
+
+Focus will be mostly on Windows, but automation will be written to be platform agnostic. We will work to introduce higher level cross-platform automation eventually. Unit tests will always be the most abundant contributor to Atom automation and should be platform-agnostic.
+
+We will accomplish this by relying as much as we can on unit tests and utilize Automated Review (AR) through LY\_Test\_Tools (LYTT) for integration tests. By focusing on O3DE wide tooling, we will avoid reinventing work that has already been done, and we will increase visibility to the wider project, since reporting and understanding will be standardized. Another benefit being simpler integration with non-Atom teams processes; our tests will be runnable without novel configuration. Regression automation will allow contributors to maintain Atom render code quality while contributing to O3DE.
```
Comment (jromnoa) [inline testing-guidance/Atom-Automation-Strategy.md:68]: i think its just "LyTestTools" without the underscores

File: testing-guidance/Atom-Automation-Strategy.md [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+*   Implement solution
+*   PR could include a test artifact file that we create a precheck github action for, although this could make Pull Requests more complicated to undertake.
+
+#### Create sample prefabs that cover the test scenarios of major render components
+
+Creating sample prefabs which lay out major render components with various options enabled will permit more rapid regression testing and provide a target for future screen compare testing.
+
+#### Convert AtomSampleViewer feature coverage to AutomatedTesting editor project
+
+Existing samples and test content in ASV should transition to AutomatedTesting where it is more accessible to customers and where it can be added to periodic or gated AR.
```
Comment (jromnoa) [inline testing-guidance/Atom-Automation-Strategy.md:104]: Should we include language here that we want to move away from using AtomSampleViewer in the future? not sure if that talk is still in process or confirmed

File: testing-guidance/Atom-Automation-Strategy.md [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+
+We want to insist on highest standards, but balance that with frugality and deliver results since unlimited time and resources are not available. Investing in the most cost efficient unit tests will reap the highest return. Maintaining a healthy level of integration tests will achieve confidence in full stack systems but should not be at the expense of unit tests.
+
+  
+
+![The Inverted Pyramid](/images/test_automation_pyramid_inverted.png)
```
Comment (jromnoa) [inline testing-guidance/Atom-Automation-Strategy.md:141]: image link doesn't work

File: testing-guidance/Atom-Automation-Strategy.md [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+
+Ideally we would have this integrated with build pipeline.
+
+OpenCPPCoverage can be used by individual engineers.
+
+Code Coverage obtained manually is a critical source of gap analysis to identify where to add tests.
+
+Developers should install OpenCPPCoverage visual studio plugin and check coverage for their tests to identify and augment gaps. Developers should include information about code coverage in code reviews.
+
+![OpenCPPCoverage VS2019 plugin](/images/OpenCPPCoverage.png)
```
Comment (jromnoa) [inline testing-guidance/Atom-Automation-Strategy.md:189]: also broken image link

File: testing-guidance/Atom-Automation-Strategy.md [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+
+*   [Using AzTest](https://www.o3de.org/docs/user-guide/testing/aztest/aztest/)
+
+### CMake/Ctest
+
+As part of AR, LYTT tests suites are declared in CmakeLists.txt files for discovery by `ctest\_entrypoint.cmd`. The ctest suite will specify a python file containing the suite of tests to run. Additional metadata for resources and dependencies are specified in the CmakeLists.txt.
+
+*   [Getting Started with Test Automation](https://www.o3de.org/docs/user-guide/testing/getting-started/)
+
+### Python LY\_Test\_Tools (LYTT pytest)
```
Comment (jromnoa) [inline testing-guidance/Atom-Automation-Strategy.md:284]: should be LyTestTools, underscores aren't needed

**Review by jromnoa — CHANGES_REQUESTED** [review]

File: testing-guidance/Atom-Automation-Strategy.md [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+    *   15.1[Q: Is there more information about this fascinating testing pyramid?](#AtomAutomationStrategy-Q:Istheremoreinformationaboutthisfascinatingtestingpyramid?)
+
+  
+
+Priority of Implementation
+--------------------------
+
+The three pillars of general availability are stability, performance, and ease of use. Automation will be scoped exclusively on stability and performance. Priority is not an assignment of ownership simply a balance of precedence.
+
+This section provides an overview of our prioritization logic based on the [Test Automation Pyramid](#AtomAutomationStrategy-TestAutomationPyramid) methodology.
```
[the remark this answers is not in the document]
Comment (smurly) [reply testing-guidance/Atom-Automation-Strategy.md:64]: these are images and lfs isn't enabled for this repo. I have to add them with a web editor to github somewhere

File: testing-guidance/Atom-Automation-Strategy.md (hunk 2 above)
Comment (smurly) [reply testing-guidance/Atom-Automation-Strategy.md:104]: They want to retain AtomSampleViewer for development work, so this is really just testing moves to AutomatedTesting project and we don't use ASV for testing

File: testing-guidance/Atom-Automation-Strategy.md (hunk 3 above)
Comment (smurly) [reply testing-guidance/Atom-Automation-Strategy.md:141]: lfs not available I'll have to figure out how to get these on github

File: testing-guidance/Atom-Automation-Strategy.md [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+Priority: Highest
+
+Owner: O3DE outside of Atom
+
+A required check for merge of github PR. This system builds and runs an automation suite that gates code. On the PR page this is labeled `continuous-integration/jenkins/pr-merge` 
+
+*   [Git Workflow](https://www.o3de.org/docs/contributing/to-code/git-workflow/)
+
+### Unit/Integration Testing Using AZTest
+
```
Comment (NeilGitGud) [inline testing-guidance/Atom-Automation-Strategy.md:271]: Missing priority

File: testing-guidance/Atom-Automation-Strategy.md [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+True unit tests are needed for platform testing. Integration "unit" tests are also using these unit test frameworks, but calling multiple modules and using live external dependencies.
+
+It is expected that unit tests will achieve 80% code coverage. Preference is for true unit tests because platform testing may only include true unit tests meaning that integration tests may never be run on non-windows.
+
+*   [Using AzTest](https://www.o3de.org/docs/user-guide/testing/aztest/aztest/)
+
+### CMake/Ctest
+
```
Comment (NeilGitGud) [inline testing-guidance/Atom-Automation-Strategy.md:279]: Missing priority

File: testing-guidance/Atom-Automation-Strategy.md [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+Priority: High (used by other methods for inclusion in AR and other systems)
+
+Owner: O3DE outside of Atom
+
+LyTestTools is a pytest based framework for running tests. This can manage fixtures and launch application under test. This provides common functionality for path resolution and other utilities used by other systems.
+
+*   [Getting Started with LyTestTools](https://www.o3de.org/docs/user-guide/testing/lytesttools/getting-started/)
+
+### Multitest Framework (pytest related)
+
```
Comment (NeilGitGud) [inline testing-guidance/Atom-Automation-Strategy.md:295]: Missing priority

File: testing-guidance/Atom-Automation-Strategy.md [hunk 10]
```diff
@@ -0,0 +1,436 @@
+  
+
+The objective of automation is the robust detection of regression early, allowing correction before issues integrate with other systems. If we allow regressions to persist, consumers of our code may inadvertently code to them, causing a negative outcome to the quality of the product. An ancillary objective of automation is the detection of previously unknown defects during the initial creation of automation.
+
+Automation supports higher confidence in the quality of our product. Code coverage is a measurable proxy for confidence in code quality. While code coverage does not equate to code quality, the level of coverage correlates to perceived confidence in quality. Our atom editor unit test code coverage is below 42% with some RHI builder dll's being below 12%; the common industry objective being greater than 80%. All software has defects; given that we are not covering our code, it is likely that defects exist in the uncovered areas.
+
```
Comment (NeilGitGud) [inline testing-guidance/Atom-Automation-Strategy.md:6]: All of these links do nothing for me, are they supposed to jump to the specified section of the document?

**Review by NeilGitGud — CHANGES_REQUESTED** [review]

File: testing-guidance/Atom-Automation-Strategy.md [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+*   Unit test creation training
+*   O3DE community documentation of how to and expectations of inclusion with contributions
+*   Unit test code coverage documentation and guidance
+
+#### Implement robust performance metric collection as an automated process
+
+This effort is already underway in service of goals established by EOY 2022. New metrics goals should be established with automatable means of collecting reproducible numbers to monitor progress.
+
+#### Expand support for batched test case patterns to include more tooling than just Editor
```
[the remark this answers is not in the document]
Comment (smurly) [reply testing-guidance/Atom-Automation-Strategy.md:86]: added more direct mention of Mulittest Framework

**Review by NeilGitGud — APPROVED** [review]

**Review by Broganab — APPROVED** [review]

**Review by jromnoa — APPROVED** [review]

**Pull request merged**